src/theory/arith/nl/equality_substitution.h

```diff
@@ -0,0 +1,87 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Gereon Kremer
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * CAD-based solver based on https://arxiv.org/pdf/2003.05633.pdf.
+ */
+
+#ifndef CVC5__THEORY__ARITH__NL__EQUALITY_SUBSTITUTION_H
+#define CVC5__THEORY__ARITH__NL__EQUALITY_SUBSTITUTION_H
+
+#include <vector>
+
+#include "context/context.h"
+#include "expr/node.h"
+#include "expr/node_algorithm.h"
+#include "smt/env_obj.h"
+#include "theory/substitutions.h"
+#include "theory/theory.h"
+
+namespace cvc5 {
+namespace theory {
+namespace arith {
+namespace nl {
+
+/**
+ * This class is a general utility to eliminate variables from a set of
+ * assertions.
+ */
+class EqualitySubstitution : protected EnvObj
+{
+ public:
+  EqualitySubstitution(Env& env);
+  /** Reset this object */
+  void reset();
+
+  /** Eliminate variables using equalities from the set of assertions */
+  std::vector<Node> eliminateEqualities(const std::vector<Node>& assertions);
+  /** Return the substitutions that were found */
+  const SubstitutionMap& getSubstitutions() const { return *d_substitutions; }
+  /** Check whether a direct conflict was found */
```

Comment:
I assume this is intended to be called immediately after eliminateEqualities? Document when this should be called.